src/pool/srv_util.c

```diff
@@ -529,3 +531,257 @@ int ds_pool_get_failed_tgt_idx(const uuid_t pool_uuid, int **failed_tgts,
+		D_DEBUG(DB_MGMT, DF_UUID": Pool cache not found\n",
+			DP_UUID(pool_id));
+		return -DER_NOSYS;
+	}
+
+	nr_downout = nr_down = 0;
+	ABT_rwlock_rdlock(pool->sp_lock);
+	for (i = 0; i < tgt_cnt; i++) {
+		nr = pool_map_find_target_by_rank_idx(pool->sp_map, rank,
+						      tgt_ids[i], &target);
+		if (nr != 1) {
+			D_ERROR(DF_UUID": Failed to get rank:%u, idx:%d\n",
+				DP_UUID(pool_id), rank, tgt_ids[i]);
+			rc = -DER_NONEXIST;
+			break;
+		}
+
+		D_ASSERT(target != NULL);
+		switch (target->ta_comp.co_status) {
+		case PO_COMP_ST_DOWNOUT:
+			nr_downout++;
+			break;
+		case PO_COMP_ST_DOWN:
+			nr_down++;
+			break;
+		default:
+			break;
+		}
+	}
+	D_ASSERT(nr_downout + nr_down <= tgt_cnt);
+
+	if (pool->sp_iv_ns != NULL)
+		*pl_rank = pool->sp_iv_ns->iv_master_rank;
+	else
+		*pl_rank = -1;
+
+	ABT_rwlock_unlock(pool->sp_lock);
+	ds_pool_put(pool);
+
+	if (rc)
+		return rc;
+	else if (nr_downout == tgt_cnt)
+		return 0;
+	else if (nr_downout + nr_down == tgt_cnt)
+		return 1;
+	else
+		return (*pl_rank == -1) ? -DER_NOSYS : 2;
+}
+
+struct exclude_targets_arg {
+	uuid_t		 eta_pool_id;
+	d_rank_t	 eta_pl_rank;
+	d_rank_t	*eta_ranks;
+	int		*eta_tgts;
+	int		 eta_nr;
+};
+
+static void
+free_exclude_targets_arg(struct exclude_targets_arg *eta)
+{
+	D_ASSERT(eta != NULL);
+	if (eta->eta_ranks != NULL)
+		D_FREE(eta->eta_ranks);
+	if (eta->eta_tgts != NULL)
+		D_FREE(eta->eta_tgts);
+	D_FREE(eta);
+}
+
+static struct exclude_targets_arg *
+alloc_exclude_targets_arg(uuid_t pool_id, int *tgt_ids, int tgt_cnt,
+			  d_rank_t pl_rank)
+{
+	struct exclude_targets_arg	*eta;
+	d_rank_t			 rank;
+	int				 i;
+
+	D_ASSERT(tgt_cnt > 0);
+	D_ASSERT(tgt_ids != NULL);
+
+	D_ALLOC_PTR(eta);
+	if (eta == NULL)
+		return NULL;
+
+	D_ALLOC_ARRAY(eta->eta_ranks, tgt_cnt);
+	if (eta->eta_ranks == NULL)
+		goto free;
+
+	D_ALLOC_ARRAY(eta->eta_tgts, tgt_cnt);
+	if (eta->eta_tgts == NULL)
+		goto free;
+
+	uuid_copy(eta->eta_pool_id, pool_id);
+	eta->eta_pl_rank = pl_rank;
+	eta->eta_nr = tgt_cnt;
+	rank = dss_self_rank();
+	for (i = 0; i < eta->eta_nr; i++) {
+		eta->eta_ranks[i] = rank;
+		eta->eta_tgts[i] = tgt_ids[i];
+	}
+
+	return eta;
+free:
+	free_exclude_targets_arg(eta);
+	return NULL;
+}
+
+static void
+exclude_targets_ult(void *arg)
+{
+	struct exclude_targets_arg	*eta = arg;
+	struct d_tgt_list		 tgt_list;
+	d_rank_list_t			 svc;
+	int				 rc;
+
+	svc.rl_ranks = &eta->eta_pl_rank;
+	svc.rl_nr = 1;
+
+	tgt_list.tl_nr = eta->eta_nr;
+	tgt_list.tl_ranks = eta->eta_ranks;
+	tgt_list.tl_tgts = eta->eta_tgts;
+
+	rc = dsc_pool_tgt_exclude(eta->eta_pool_id, NULL /* grp */, &svc,
+				  &tgt_list);
+	if (rc)
+		D_ERROR(DF_UUID": Exclude targets failed. %d\n",
+			DP_UUID(eta->eta_pool_id), rc);
+
+	free_exclude_targets_arg(eta);
+}
+
+/*
+ * The NVMe faulty reaction is called from bio_nvme_poll() which is on
+ * progress (hardware poll) ULT, and it will call into client stack to
+ * exclude pool targets, blocking calls could be made in this code path,
```

Comment:
"Could not be"?

Reply:
Yeah, the comment is bit confusing. I meant the client API code path may have blocking calls, so we can't directly call it on this progress ULT.